File: 22. 括号生成.js
/**
 * @param {number} n
 * @return {string[]}
 */
 var generateParenthesis = function(n) {
    var result = []
    var DFS = function(str, n, nowLeft, nowRight) {
        if (str.length == n * 2) {
            result.push(str);
            return;
        }
        if (nowLeft < n) DFS(str + '(', n, nowLeft + 1, nowRight);
        if (nowRight < nowLeft) DFS(str + ')', n, nowLeft, nowRight + 1);
    }
    DFS('', n, 0, 0);
    return result;
};

File: 22. 括号生成.cpp
class Solution {
public:
    vector<string> v;
    void DFS(int left, int right, int n, string s) {
        if (left == n && right == n) {
            v.push_back(s);
            return;
        }
        if (left < n) DFS(left + 1, right, n, s + "(");
        if (right < left) DFS(left, right + 1, n , s + ")");
    }
    vector<string> generateParenthesis(int n) {
        DFS(0, 0, n, "");
        return v;
    }
};